### PR TITLE
Feature: Add Ruff GitHub Workflow

### PR DESCRIPTION
**Pull request by error9098x**
Features:
- Adds a GitHub workflow for Ruff linter for better code quality.
- Automatically runs after every PR.
- Automatically tries to fix the issues and raise another PR with fixed issues.
- If unable to fix the issues, the workflow fails

Screenshot from 2024-06-22 14-42-18

Screenshot from 2024-06-22 14-43-06

### CODE REVIEW

**Comment by luarss** [comment]
Can you add some docs for this?

File: .github/workflows/ruff_linter.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    - name: Check if Ruff Found Errors
+      id: check-ruff-errors
+      run: |
+        if [[ "${{ env.RUFF_OUTPUT }}" != *"All checks passed!"* ]]; then
+          echo "Errors found by Ruff."
+          echo "errors_found=true" >> $GITHUB_OUTPUT
+        else
+          echo "No errors found by Ruff."
+          echo "errors_found=false" >> $GITHUB_OUTPUT
+        fi
```
Comment (luarss) [inline .github/workflows/ruff_linter.yml:49]: Does the CI fail if format checks are failing? It seems it is just printing stuff.

File: .github/workflows/ruff_linter.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      uses: peter-evans/create-pull-request@v5
+      with:
+        title: 'style fixes by ruff'
+        body: 'This PR contains automated style fixes generated by Ruff.'
+        branch: 'ruff/style-fixes'
+        labels: 'auto-generated'
+        commit-message: 'style fixes by ruff'
+        author: 'github-actions[bot] <github-actions[bot]@users.noreply.github.com>'
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (luarss) [inline .github/workflows/ruff_linter.yml:69]:
```suggestion
```

Please avoid creating a PR for passing lint tests.

File: .github/workflows/ruff_linter.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
+name: Ruff Linter and Auto Fix
+
+on: [push, pull_request]
```
Comment (luarss) [inline .github/workflows/ruff_linter.yml:3]:
```suggestion

on:
  push:
    branches:
      - feature/add-ruff-github-workflow # TODO: remove.
      - master
  pull_request:
    branches:
      - feature/add-ruff-github-workflow # TODO: remove.
      - master
```

Comment (luarss) [reply .github/workflows/ruff_linter.yml:3]: This enables testing of actions in this repo.

File: .github/workflows/ruff_linter.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: 3.12
+
+    - name: Install Ruff
+      run: |
+        python -m pip install --upgrade pip
+        pip install ruff
```
Comment (vvbandeira) [inline .github/workflows/ruff_linter.yml:21]: Pinning a version of Python libs (i.e., `ruff==x.x.x`) would be a good idea. As well as regularly updating them, this way we can isolate and understand the difference between lib versions.

**Review by vvbandeira — COMMENTED** [review]
I am just commenting on the GH workflow. In the future, keep PRs single-focused, either CI-related or related to code format or code changes. Also, let me know if GHA VMs meet the required HW resources or when you would like to set up an automatic deployment of the front/backend. This will require some dev time from me, so let me know in advance so I can plan accordingly.

**Comment by luarss** [comment]
@vvbandeira Understood, it's just that this is the very first addition of the CI, so it warrants some code/requirement changes.

**Review by vvbandeira — DISMISSED** [review]
LGTM

**Review by vvbandeira — DISMISSED** [review]

**Review by vvbandeira — APPROVED** [review]

**Pull request merged**